### PR TITLE
Update signature verification

### PR DESCRIPTION
**Pull request by tenkod**
[`StrongNameSignatureVerificationEx`](https://msdn.microsoft.com/en-us/library/ms232579) has been obsoleted.

Use `ICLRStrongName.StrongNameSignatureVerificationEx` instead:

``` c#
/// <summary>
/// Invokes <c>ICLRStrongName.StrongNameSignatureVerificationEx</c>
/// (see https://msdn.microsoft.com/en-us/library/ff844054 for more info).
/// </summary>
/// <remarks>
/// If we ever need failure details, <c>Microsoft.Runtime.Hosting.StrongNameHelpers.StrongNameErrorInfo</c>
/// method can be used immediately after the verify call to get the HRESULT;
/// more information about .NET CLR HRESULT codes can be found here:
/// http://blogs.msdn.com/b/yizhang/archive/2010/12/17/interpreting-hresults-returned-from-net-clr-0x8013xxxx.aspx
/// </remarks>
public static bool VerifySignature(string assemblyPath, bool forceCheck = false)
{
    bool wasVerified;
    return VerifySignatureMethod(assemblyPath, forceCheck, out wasVerified) && wasVerified;
}


private delegate bool VerifySignatureDelegate(string assemblyPath, bool forceCheck, out bool wasVerified);

private static readonly VerifySignatureDelegate VerifySignatureMethod = LoadVerifySignatureMethod();

private static VerifySignatureDelegate LoadVerifySignatureMethod()
{
    var strongNameHelpers = typeof (AppDomain).Assembly.GetType("Microsoft.Runtime.Hosting.StrongNameHelpers");
    if (strongNameHelpers == null) { throw new InvalidOperationException("StrongNameHelpers cannot be found in mscorlib."); }
    var verifySignatureMethod = strongNameHelpers.GetMethod("StrongNameSignatureVerificationEx");
    if (verifySignatureMethod == null) { throw new InvalidOperationException("StrongNameHelpers.StrongNameSignatureVerificationEx cannot be found in mscorlib."); }
    return (VerifySignatureDelegate) verifySignatureMethod.CreateDelegate(typeof (VerifySignatureDelegate));
}
```

### CODE REVIEW

**Comment by tenkod** [comment]
@milang , please review.

File: src/Pandell.PackageRunner/AssemblyChecker.cs [hunk 1]
```diff
@@ -9,17 +8,33 @@ namespace PackageRunner
     /// </summary>
     internal class AssemblyChecker
     {
-        [DllImport("mscoree.dll", CharSet = CharSet.Unicode)]
-        private static extern bool StrongNameSignatureVerificationEx(string wszFilePath, byte fForceVerification, ref byte pfWasVerified);
+        private delegate bool VerifySignatureDelegate(string assemblyPath, bool forceCheck, out bool wasVerified);
```
Comment (milang) [inline src/Pandell.PackageRunner/AssemblyChecker.cs]: Please move the `private` members to the bottom of the class, after a [private separator](https://github.com/pandell/Pli/blob/v4.6.0.213/Source/Pandell.Common/ObjectExtensions.cs#L99-L101).

**Comment by milang** [comment]
Would it be possible to write a unit test that runs `IsStrongNameValid` against one of the signed system assemblies and another test that runs it against an unsigned assembly? (to get test for success and failure).

**Comment by milang** [comment]
Finished this round.

**Comment by tenkod** [comment]
@milang , changes applied and tests added. Please have another look.

**Comment by milang** [comment]
@tenkod please note that I pushed some changes to this branch, you will have to fetch & reset on your machine (I also recommend clobber + build).

**Comment by milang** [comment]
Finished my changes. @tenkod , @pandell/core-developers please review.

File: src/Pandell.PackageRunner/AssemblyExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+                        return Assembly.Load(assemblyBytes);
+                    }
+                })
+                .ToArray();
+
+            AppDomain.CurrentDomain.AssemblyResolve += (sender, eventArgs) =>
+            {
+                var assemblyToResolve = new AssemblyName(eventArgs.Name);
+                return embeddedAssemblies.FirstOrDefault(
+                    a => AssemblyName.ReferenceMatchesDefinition(assemblyToResolve, a.GetName()));
```
Comment (tenkod) [inline src/Pandell.PackageRunner/AssemblyExtensions.cs:51]: It looks like that `ReferenceMatchesDefinition` has a flaw and compares only the names:
From MSDN (https://msdn.microsoft.com/en-us/library/system.reflection.assemblyname.referencematchesdefinition(v=vs.110).aspx#Anchor_1): 

> The comparison depends only on the simple assembly name. It ignores version, culture, and public key token.

And also here they said that they won't fix it to not break existing apps: https://connect.microsoft.com/VisualStudio/feedback/details/752902/assemblyname-referencematchesdefinition-returns-true-even-though-the-parameters-dont-match
Should we revert back to full name string compare?

**Comment by tenkod** [comment]
@milang , changes looks good to me, thanks! Had just a one minor question/comment.

File: src/Pandell.PackageRunner/AssemblyExtensions.cs (hunk 2 above)
Comment (tenkod) [inline src/Pandell.PackageRunner/AssemblyExtensions.cs:51]: @milang , are you ok if we replace this with full name string compare and merge it?

Comment (milang) [inline src/Pandell.PackageRunner/AssemblyExtensions.cs:51]: No, I am not ok with replacing with full name comparison.

Comment (tenkod) [inline src/Pandell.PackageRunner/AssemblyExtensions.cs:51]: Ok, then merging it.

**Pull request merged**